File: DesignPatterns/01_GOF23/DP_examples/Lua/02_Structural/Facade/Subsystem.lua
-- 形状接口
---@class IShape
---@field draw fun()
---@field new fun(o?: table)
IShape = {}
function IShape:draw()
    error("Subclass must implement draw method")
end
function IShape:new(o)
    o = o or {}
    setmetatable(o, self)
    self.__index = self
    return o
end

-- 圆形实现
---@class Circle : IShape
---@field radius number
Circle = IShape:new()
function Circle:new(radius)
    local o = IShape.new(self)
    o.radius = radius
    return o
end
function Circle:draw()
    print(string.format("Draw a circle with radius %d", self.radius))
end

-- 矩形实现
---@class Rectangle : IShape
---@field width number
---@field height number
Rectangle = IShape:new()
function Rectangle:new(width, height)
    local o = IShape.new(self)
    o.width = width
    o.height = height
    return o
end
function Rectangle:draw()
    print(string.format("Draw a rectangle with width %d and height %d", self.width, self.height))
end

-- 正方形实现
---@class Square : IShape
---@field side number
Square = IShape:new()
function Square:new(side)
    local o = IShape.new(self)
    o.side = side
    return o
end
function Square:draw()
    print(string.format("Draw a square with side %d", self.side))
end

File: DesignPatterns/01_GOF23/DP_examples/Python/02_Structural/Facade/Subsystem.py
from abc import ABC, abstractmethod

# 形状接口
class IShape(ABC):
    @abstractmethod
    def draw(self):
        pass

# 圆形实现
class Circle(IShape):
    def __init__(self, radius):
        self.radius = radius
    
    def draw(self):
        print(f"Draw a circle with radius {self.radius}")

# 矩形实现
class Rectangle(IShape):
    def __init__(self, width, height):
        self.width = width
        self.height = height
    
    def draw(self):
        print(f"Draw a rectangle with width {self.width} and height {self.height}")

# 正方形实现
class Square(IShape):
    def __init__(self, side):
        self.side = side
    
    def draw(self):
        print(f"Draw a square with side {self.side}")
